### PR TITLE
[BOSK-60] Global mechanism to set a random state for blocks and pipelines

### PR DESCRIPTION
**Pull request by Stasychbr**
Some blocks in the `zoo` were adjusted and some functionality was added for the `auto_block` and `BasePipeline`. Tests were corrected according to the made changes.

### CODE REVIEW

File: bosk/block/auto.py [hunk 1]
```diff
@@ -127,6 +132,17 @@ def __setstate__(self, state: dict):
...
 
+            def set_random_state(self, seed: int) -> None:
+                if hasattr(self.__instance, 'set_random_state'):
+                    return self.__instance.set_random_state(seed)
+                if random_state_field is None:
+                    return super().set_random_state(seed)
+                if hasattr(self.__instance, random_state_field):
+                    setattr(self.__instance, random_state_field, seed)
+                else:
+                    logging.warning("%s doesn't have random_state_field '%s'", cls.__name__,
```
Comment (andruekonst) [inline bosk/block/auto.py:143]: Please, use the python builtin `warning` lib instead of `logging.warning`, because it is more flexible. Warnings will be automatically intercepted by `logging` if it is used, otherwise they will be written to stderr.

File: bosk/block/zoo/data_weighting/data_weighting.py [hunk 2]
```diff
@@ -8,7 +8,9 @@
 from bosk.data import CPUData, GPUData
 
 
-@auto_block(auto_state=True, execution_props=BlockExecutionProperties(cpu=True, gpu=True))
+@auto_block(auto_state=True,
+            execution_props=BlockExecutionProperties(cpu=True, gpu=True),
```
Comment (andruekonst) [inline bosk/block/zoo/data_weighting/data_weighting.py:12]: Not relevant for exactly this pull request, but `gpu=True` seems like wrong default behavior, isn't it @AndreyAgeev?

File: bosk/block/zoo/multi_grained_scanning/multi_grained_scanning.py [hunk 3]
```diff
@@ -49,3 +50,11 @@ def _window_slicing_predict(self, X) -> 'np.ndarray':
 
     def transform(self, X) -> CPUData:
         return CPUData(self._window_slicing_predict(X))
+
+    def set_random_state(self, seed: int) -> None:
+        for model in self._models:
+            if hasattr(model, 'random_state'):
+                model.random_state = seed
```
Comment (andruekonst) [inline bosk/block/zoo/multi_grained_scanning/multi_grained_scanning.py:57]: Probably, setting the same random seed for each model would lead to undesired results. It is expected, that the base models are always different, so the random seeds should be different too.

Example of handling such a case can be found [here, in the sklearn `BaggingEstimator`](https://github.com/scikit-learn/scikit-learn/blob/9aaed498795f68e5956ea762fef9c440ca9eb239/sklearn/ensemble/_bagging.py#L469): instead of using the same value, the seed can be used to populate seeds for the underlying models.

File: bosk/pipeline/base.py [hunk 4]
```diff
@@ -39,3 +39,8 @@ def accept(self, visitor: BaseVisitor):
             conn.accept(visitor)
 
         visitor.visit(self)
+
+    def set_random_state(self, seed: int) -> None:
+        """Set random seed for each block in the pipeline."""
+        for block in self.nodes:
+            block.set_random_state(seed)
```
Comment (andruekonst) [inline bosk/pipeline/base.py:46]: Here is the same problem as in `multi_grained_scanning.py`.
For example, imagine a pipeline consisting of two parallel random forests: if we will use the same random seed, the forests will be identical which won't improve the ensemble accuracy.
To avoid it populate separate random seed for each node, based on the given seed.

**Review by andruekonst — CHANGES_REQUESTED** [review]
All is ok, except how seeds are set for models that consist of underlying models.

File: bosk/pipeline/base.py [hunk 5]
```diff
@@ -39,3 +41,9 @@ def accept(self, visitor: BaseVisitor):
             conn.accept(visitor)
 
         visitor.visit(self)
+
+    def set_random_state(self, seed: int | Generator) -> None:
+        """Set random seed for each block in the pipeline."""
+        gen = get_random_generator(seed)
+        for block in self.nodes:
+            block.set_random_state(gen)
```
Comment (andruekonst) [inline bosk/pipeline/base.py:49]: Using the same random generator in different blocks may result in irreproducibility in case of parallel pipeline execution.
I think that it is better to generate seeds here.

**Review by andruekonst — APPROVED** [review]
